conn_browser.go

```diff
@@ -0,0 +1,229 @@
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+	}
+	return nil
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	return NewAddr("0.0.0.0:0")
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	rawURL := c.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) SetDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetReadDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) SetWriteDeadline(t time.Time) error {
+	return nil
+}
+
+func (c *Conn) setUpHandlers() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		// Message handlers already created. Nothing to do.
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			// TODO(albrow): Currently we assume data is of type Blob. Really we
+			// should check binaryType and then decode accordingly.
+			blob := args[0].Get("data")
+			data := readBlob(blob)
+			c.incomingData <- data
+		}()
+		return nil
+	})
+	c.messageHandler = &messageHandler
+	c.Call("addEventListener", "message", messageHandler)
+
+	closeHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		close(c.closeSignal)
+		return nil
+	})
+	c.closeHandler = &closeHandler
+	c.Call("addEventListener", "close", closeHandler)
+}
+
+// readLoop continuosly reads from the c.incoming data channel and writes to the
+// current data buffer.
+func (c *Conn) readLoop() error {
+	for data := range c.incomingData {
+		c.currDataMut.Lock()
+		_, err := c.currData.Write(data)
+		if err != nil {
+			c.currDataMut.Unlock()
+			return err
+		}
+		c.currDataMut.Unlock()
+		c.dataSignal <- struct{}{}
+	}
+	return nil
+}
+
+func (c *Conn) waitForOpen() error {
+	openSignal := make(chan struct{})
+	handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		close(openSignal)
+		return nil
+	})
+	defer c.Call("removeEventListener", "open", handler)
+	defer handler.Release()
+	c.Call("addEventListener", "open", handler)
+	<-openSignal
+	return nil
+}
+
+// readBlob converts a JavaScript Blob into a slice of bytes. It uses the
+// FileReader API under the hood.
+func readBlob(blob js.Value) []byte {
+	reader := js.Global().Get("FileReader").New()
+	dataChan := make(chan []byte)
+	loadEndFunc := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		// event.result is of type ArrayBuffer. We can convert it to a JavaScript
+		// Uint8Array, which is directly translatable to the Go type []byte.
+		buffer := reader.Get("result")
+		view := js.Global().Get("Uint8Array").New(buffer)
+		dataLen := view.Get("length").Int()
+		data := make([]byte, dataLen)
+		for i := 0; i < dataLen; i++ {
+			data[i] = byte(view.Index(i).Int())
```

Comment:
This is going to be painfully slow. Is there no other way to get data out of javascript? Even if it involves allocations, getting the browser to give us something we can just read all at once would be faster.

Reply:
I just pushed a commit that switches to using the `readAsBinaryString` method of `FileReader`. Need to do a bit more testing to make sure that this works for any arbitrary data and we don't run into encoding issues. I'm somewhat worries because I have run into issues with the `binary` encoding in JavaScript in the past.

For some additional background, the `onmessage` handler in the JavaScript WebSockets API gives us data of type `Blob`. The [Blob API](https://developer.mozilla.org/en-US/docs/Web/API/Blob) in JavaScript is fairly obtuse. AFAICT the only way to read from a `Blob` is to use a [`FileReader`](https://developer.mozilla.org/en-US/docs/Web/API/FileReader). There are a few different ways to use the `FileReader` API. I originally tried using the `readAsText` method, but was running into problems with how JavaScript handles string encoding. That's why I switched to the `readAsArrayBuffer` method which turns the data into an `ArrayBuffer` which can be converted to `Uint8Array`, which is the JavaScript equivalent of `[]byte`. There is a third method that might be faster for us called `readAsBinaryString`. I think it's faster because we can read the data as a `string` all at once and then convert that to `[]byte` in Go without making anymore `syscall/js` calls.

Reply:
Unfortunately `readAsBinaryString` doesn't work. You can't convert directly from a binary string in JavaScript to `[]byte` in Go (I suspect JavaScript is doing something wonky with their "binary" encoding).

When testing out the `readAsBinaryString` method in [0x Mesh](https://github.com/0xProject/0x-mesh), I was seeing the following error:

```
failed to negotiate security protocol: unexpected EOF
```

Here's the first couple frames as reference:

Using `readAsArrayBuffer`:

```
00000000  13 2f 6d 75 6c 74 69 73  74 72 65 61 6d 2f 31 2e  |./multistream/1.|
00000010  30 2e 30 0a                                       |0.0.|

00000000  0d 2f 73 65 63 69 6f 2f  31 2e 30 2e 30 0a        |./secio/1.0.0.|

00000000  00 00 00 6c 0a 10 4e 2b  ae 04 e3 d0 11 a6 d9 6a  |...l..N+.......j|
00000010  62 a6 7c 79 6f 0a 12 25  08 02 12 21 03 3f d2 e5  |b.|yo..%...!.?..|
00000020  b5 9a 7c 00 8c 52 e7 23  25 b1 17 36 47 b3 90 89  |..|..R.#%..6G...|
00000030  6b 99 66 0e 56 b1 5c bf  84 9f 52 ee 56 1a 11 50  |k.f.V.\...R.V..P|
00000040  2d 32 35 36 2c 50 2d 33  38 34 2c 50 2d 35 32 31  |-256,P-384,P-521|
00000050  22 0f 41 45 53 2d 32 35  36 2c 41 45 53 2d 31 32  |".AES-256,AES-12|
00000060  38 2a 0d 53 48 41 32 35  36 2c 53 48 41 35 31 32  |8*.SHA256,SHA512|
```

Using `readAsBinaryString`:

```
00000000  13 2f 6d 75 6c 74 69 73  74 72 65 61 6d 2f 31 2e  |./multistream/1.|
00000010  30 2e 30 0a                                       |0.0.|

00000000  0d 2f 73 65 63 69 6f 2f  31 2e 30 2e 30 0a        |./secio/1.0.0.|

00000000  00 00 00 6c 0a 10 c3 b5  c2 b7 c3 8b c2 8d 12 c3  |...l............|
00000010  ac c2 ac c2 b7 c2 92 c3  a1 c2 9b 3a 78 36 59 6a  |...........:x6Yj|
00000020  12 25 08 02 12 21 03 3f  c3 92 c3 a5 c2 b5 c2 9a  |.%...!.?........|
00000030  7c 00 c2 8c 52 c3 a7 23  25 c2 b1 17 36 47 c2 b3  ||...R..#%...6G..|
00000040  c2 90 c2 89 6b c2 99 66  0e 56 c2 b1 5c c2 bf c2  |....k..f.V..\...|
00000050  84 c2 9f 52 c3 ae 56 1a  11 50 2d 32 35 36 2c 50  |...R..V..P-256,P|
00000060  2d 33 38 34 2c 50 2d 35  32 31 22 0f 41 45 53 2d  |-384,P-521".AES-|
00000070  32 35 36 2c 41 45 53 2d  31 32 38 2a 0d 53 48 41  |256,AES-128*.SHA|
00000080  32 35 36 2c 53 48 41 35  31 32                    |256,SHA512|
```

@Stebalien does anything in that data look weird to you? If I'm right and JavaScript is doing something weird with their binary encoding maybe there is an efficient way to correct the encoding and then convert to `[]byte`?

Otherwise, it looks like Go 1.13 is going to come with a new [`js.CopyBytesToGo` function](https://github.com/golang/go/commit/c468ad04177c422534ad1ed4547295935f84743d) that we could use with the `readAsArrayBuffer` method. So we might just have to go with slower manual copying for now and switch to that when Go 1.13 comes out.

Reply:
> Otherwise, it looks like Go 1.13 is going to come with a new js.CopyBytesToGo function that we could use with the readAsArrayBuffer method. So we might just have to go with slower manual copying for now and switch to that when Go 1.13 comes out.

Sounds good. Yeah, it looks like an encoding issue.

Reply:
@Stebalien I just pushed up a new commit that forces the JavaScript WebSockets API to use `ArrayBuffer` instead of `Blob` for the type of incoming messages. This means we don't have to use the `FileReader` API and we can avoid a lot of headaches with trying to turn asynchronous JavaScript code into synchronous Go code. However, we do still have to manually allocate a new `[]byte` and use a for loop to copy the data for now.